desc/protoparse/ast/walk.go

```diff
@@ -0,0 +1,493 @@
+package ast
+
+// VisitFunc is used to examine a node in the AST when walking the tree.
+// It returns true or false as to whether or not the descendants of the
+// given node should be visited. If it returns true, the node's children
+// will be visisted; if false, they will not. When returning true, it
+// can also return a new VisitFunc to use for the children. If it returns
+// (true, nil), then the current function will be re-used when visiting
+// the children.
+//
+// See also the Visitor type.
+type VisitFunc func(Node) (bool, VisitFunc)
+
+// Walk conducts a walk of the AST rooted at the given root using the
+// given function. It performs a "pre-order traversal", visiting a
+// given AST node before it visits that node's descendants.
+func Walk(root Node, v VisitFunc) {
+	ok, next := v(root)
+	if !ok {
+		return
+	}
+	if next != nil {
+		v = next
+	}
+	if comp, ok := root.(CompositeNode); ok {
+		for _, child := range comp.Children() {
+			Walk(child, v)
+		}
+	}
+}
+
+// Visitor provides a technique for walking the AST that allows for
+// dynamic dispatch, where a particular function is invoked based on
+// the runtime type of the argument.
+//
+// It consists of a number of functions, each of which matches a
+// concrete Node type. It also includes functions for sub-interfaces
+// of Node and the Node interface itself, to be used as broader
+// "catch all" functions.
+//
+// To use a visitor, provide a function for the node types of
+// interest and pass visitor.Visit as the function to a Walk operation.
+// When a node is traversed, the corresponding function field of
+// the visitor is invoked, if not nil. If the function for a node's
+// concrete type is nil/absent but the function for an interface it
+// implements is present, that interface visit function will be used
+// instead. If no matching function is present, the traversal will
+// continue. If a matching function is present, it will be invoked
+// and its response determines how the traversal proceeds.
+//
+// Every visit function returns (bool, *Visitor). If the bool returned
```

Comment:
`Visitor` is a struct, not an interface

If it were an interface, it would be very cumbersome to create new ones as you'd have to implement _every_ method. Instead, you just implement the relevant methods and add them to the visitor struct as function references.

This is much simpler than allowing embedding the interface + overrides, as the walk could not easily tell which methods were overridden and which were not and would have to catch nil-dereference panics to determine when the method is not actually present (which seemed very smelly).